Add the release runbook to RELEASING.md (#2533)
opensearch-project/OpenSearch-Dashboards: RELEASING.md

## Releasing

This project follows [OpenSearch project branching, labelling, and releasing](https://github.com/opensearch-project/.github/blob/main/RELEASING.md).

## Runbook

### Overview

> The OpenSearch project releases as versioned distributions of OpenSearch, OpenSearch Dashboards, and the OpenSearch plugins. This runbook describes in detail how to perform an end-to-end major/minor/patch version release for OpenSearch Dashboards project as the release manager (RM) of this release version. RM is responsible for updating the release status broadly on the release tracking issue on github.

Reviewer: `github` -> `GitHub`.

#### Important Dates

> - Next minor release: [Version 2.4.0 (November 10, 2022)](https://github.com/orgs/opensearch-project/projects/1#column-19044869)

Reviewer: NIT: I would remove these dates as this is maintained in GitHub issues and by the build repo and subject to be delayed if any security issues occur.

Reviewer 2: You can just refer to https://opensearch.org/releases.html

- Code freeze date for next minor release: [Version 2.4.0 (November 03, 2022)](https://github.com/opensearch-project/openSearch-build/issues/2649#issue-1378667303)

### Release Phase 1 - Preparation

> To track the major / minor / patch release, a github issue will be created. The issue will be created on the OpenSearch Dashboards repository for major / minor releases, with the overall issue tracked on the OpenSearch build repository. Only an overall issue will be created on the OpenSearch build repository for patch release.

Reviewer 3: ```suggestion
For major and minor releases, the build team will create a release issue in the OpenSearch Dashboards repository ([example](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2230)) which links to the overall issue in the OpenSearch build repository ([example](https://github.com/opensearch-project/opensearch-build/issues/2447)). For patch releases, they will only create the overall issue in the OpenSearch build repository ([example](https://github.com/opensearch-project/opensearch-build/issues/2650)).
```

Whenever possible, we should be specific about who is responsible for any actions.

(Example github issues of [major](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1548) / [minor](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2230) / [patch](https://github.com/opensearch-project/opensearch-build/issues/2650) release)

The release manager (RM) will be assigned to the OpenSearch Dashboards release issue and will be responsible for reviewing all tasks in this issue for preparation. It's also a good idea to compare the current release issue to the issue of previous release version at this point to ensure that all new processes have been captured.

> RM will review the [public roadmap](https://github.com/orgs/opensearch-project/projects/1) and confirm the release scope with other OpenSearch Dashboards [maintainers](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/MAINTAINERS.md) as well as feature owners. All features not ready for current release will be relabeled. For example, if current release version is v2.3.0, all features not ready will be labeled as v2.4.0. (or even later by discussing with the feature owner). After then, RM will need to check all PRs for the current release version are merged into main branch and backported PRs are merged with all CI passed.

Reviewer: `v2.4.0. (or`

i would drop the 3rd `.`

Reviewer 3: This is good info about labels, but we could emphasize that release labels are intended to clearly communicate which features and fixes are targeted for the upcoming release. So this step is basically an audit to make sure issues that should have the label have it, and those that won't make it don't.

#### Prepare BWC data and update BWC versions

> Backwards Compatibility Tests (BWC) are cypress tests that verify any changes are backwards compatible with previous versions. RM will generate test data and test locally following instructions [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/TESTING.md#backwards-compatibility-tests) and cut PR to include both generated data and version upgrade for automated build. (See example [PR](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2393/files))

Reviewer 3: > that verify any changes are backwards compatible with previous versions

This is worded a bit stronger than the reality - unfortunately, BWC tests don't guarantee compatibility, they just help us catch obvious compatibility bugs.

Author: It seems there is room for improvement 😉

#### Cut release branch for major / minor release

For major / minor release, RM will need to cut the release branch from the parent branch, [following OpenSearch project branching](https://github.com/opensearch-project/.github/blob/main/RELEASING.md#opensearch-branching)

### Release Phase 2 - Pre-Release

> The release process for OpenSearch is centralized. Jenkins workflows are in place to run daily snapshot builds for OpenSearch and OpenSearch Dashboards. RM will update the release branch version in the distribution manifest (see example [PR](https://github.com/opensearch-project/opensearch-build/pull/2586/files)) and increment the parent branch version (see example [PR](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2295/files)).

Reviewer: NIT: `daily` snapshots aren't build. As far as I am aware, and I could be wrong but the current workflow actually is a CRON job checks if there is a diff in the components. Depending on how close the release is for the version the CRON job is more frequent going from like 6 hours to like 2 hours or something. And if there aren't changes then there is no build.

### Release Phase 3 - Release testing

#### Verify integration and BWC test results

The automated integration test and BWC test are executed concurrently with the release artifacts build. RM should examine the test results and assist in triaging the broken test case.

Example build:

x64: https://build.ci.opensearch.org/job/integ-test-opensearch-dashboards/995/

arm64: https://build.ci.opensearch.org/job/integ-test-opensearch-dashboards/996/

Example test results:
https://opensearch-project.github.io/opensearch-dashboards-functional-test/site/?version=2.3.0&build_number=4104&test_number=996

Note: change `arch` to correspond to the build's processor (CPU) architecture

#### Sanity test with tarball and docker image

Once the release candidate artifacts are built, RM will configure the OpenSearch cluster with OpenSearch Dashboards according to the [instructions in the OpenSearch build repo](https://github.com/opensearch-project/opensearch-build/issues/2447#issuecomment-1241406594) and produce sanity tests to identify broken functionalities caused by new features / code changes. If you find any, please file bug reports and assist in triaging the bugfix.

> ### Release Phase 4 - Release Announcement

Reviewer: Not sure if we include that after the release announced the build repo team will trigger a job that will cut a tag in each repo from the commit hash and branch that was included in the release (which could take a couple hours for it to be available). But it is on the maintainers of the repo to create the release form the tag automatically created.

Author: Yes i didn't put the tag generate part since it's not owned by OSD, but I can add it here.

The second part is included in Post-Release session.

## Releasing

This project follows [OpenSearch project branching, labelling, and releasing](https://github.com/opensearch-project/.github/blob/main/RELEASING.md).

## Runbook

### Overview

> The OpenSearch project releases versioned distributions of OpenSearch, OpenSearch Dashboards, and the OpenSearch plugins. This runbook describes in detail how to perform an end-to-end major/minor/patch version release for the OpenSearch Dashboards project as the release manager (RM) of this release version. RM is responsible for updating the release status broadly on the release tracking issue ([example](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2230)) on GitHub.

Reviewer 4: ```suggestion
The OpenSearch project releases versioned distributions of OpenSearch, OpenSearch Dashboards, and the OpenSearch plugins. This runbook details the steps involved in performing major, minor, and patch version releases for the OpenSearch Dashboards project; these steps need to be completed by the release managers (RM) assigned to each release. The RM is also responsible for updating the release status on the release tracking issues maintained on GitHub ([example](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2230)).
```

\*Important Dates: https://opensearch.org/releases.html

### Release Phase 1 - Preparation

For major and minor releases, the OpenSearch build repository [maintainers](https://github.com/opensearch-project/opensearch-build/blob/main/MAINTAINERS.md) will create a release issue in the OpenSearch Dashboards repository ([example](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2230)) which links to the overall issue in the OpenSearch build repository ([example](https://github.com/opensearch-project/opensearch-build/issues/2447)). For patch releases, they will only create the overall issue in the OpenSearch build repository ([example](https://github.com/opensearch-project/opensearch-build/issues/2650)).

> The release manager (RM) will be assigned to the OpenSearch Dashboards release issue and will be responsible for reviewing all tasks in this issue for preparation. It's also a good idea to compare the current release issue to the issue of previous release version at this point to ensure that all new processes have been captured.

Reviewer 4: ```suggestion
The OpenSearch Dashboards release issue will be assigned to the RM who will prepare for the release by reviewing all listed tasks. They should also compare the current release issue to the issue of the previous release version to ensure that all new processes have been captured.
```

also, what is a process in `all new processes`?

Author: The release process is evolving with efforts from contributors. Couple of examples - 1. build pipeline will create the release tag automatically instead of manually 2. RM can create release notes based on CHANGELOG in v2.4.0

> RM will review the [public roadmap](https://github.com/orgs/opensearch-project/projects/1) and confirm the release scope with other OpenSearch Dashboards [maintainers](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/MAINTAINERS.md) as well as feature owners. Since release labels are intended to clearly communicate which features and fixes are targeted for the upcoming release, RM will audit that all issues and pull requests are labelled properly. For example, if current release version is v2.3.0, all features not ready will be labeled as v2.4.0 (or even later by discussing with the feature owner). After then, RM will need to check all PRs for the current release version are merged into main branch and backported PRs are merged with all CI passed.

Reviewer 4: ```suggestion
The RM will review the [public roadmap](https://github.com/orgs/opensearch-project/projects/1) and confirm the release scope with other OpenSearch Dashboards [maintainers](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/MAINTAINERS.md) as well as the feature owners. Since release labels are intended to highlight the features and fixes meant for the upcoming release, the RM will verify that all issues and pull requests are labelled accordingly. For example, if current release version was v2.3.0, all features not ready for the release should be labeled as v2.4.0 or later by discussing with the feature owners. The RM will also need to check all PRs for the current release version to confirm that they are merged into the `main` branch and their backported PRs are merged with all CI passing.
```
